src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilerDiagnosticAnalyzer.CompilationAnalyzer.cs

```diff
@@ -116,7 +120,11 @@ public override int GetHashCode()
 
                 public override bool Equals(Diagnostic? obj)
                 {
-                    return _original.Equals(obj);
+                    if (obj instanceof CompilerDiagnostic cd) {
+                        return _original.Equals(cd._original);
+                    } else {
+                        return _original.Equals(obj);
+                    }
```

Comment:
something feels wrong about this fix to me.  i'd also like to see a repro/test.  calling through _original.Equals(obj) should be correct.  If not, something about hte underlying equals isn't properly handling things.

Reply:
@CyrusNajmabadi It 100% does not work here.  The repro is really trivial:  Create an instance of this class (passing in a different Diagnostic, of course).  Then call `a.Equals(a)`. This will return false.

This is because `a.original` does not equal `a`.